lib/livebook/hubs/fly.ex

```diff
@@ -1,31 +1,95 @@
 defmodule Livebook.Hubs.Fly do
   @moduledoc false
-  defstruct [
-    :id,
-    :access_token,
-    :hub_name,
-    :hub_color,
-    :organization_id,
-    :organization_type,
-    :organization_name,
-    :application_id
-  ]
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Livebook.{Hubs, Utils}
 
   @type t :: %__MODULE__{
           id: Livebook.Utils.id(),
           access_token: String.t(),
           hub_name: String.t(),
-          hub_color: Livebook.Users.User.hex_color(),
+          hub_color: String.t(),
           organization_id: String.t(),
           organization_type: String.t(),
           organization_name: String.t(),
           application_id: String.t()
         }
 
-  def save_fly(fly, params) do
-    fly = %{fly | hub_name: params["hub_name"], hub_color: params["hub_color"]}
+  embedded_schema do
+    field(:access_token, :string)
+    field(:hub_name, :string)
+    field(:hub_color, :string)
+    field(:organization_id, :string)
+    field(:organization_type, :string)
+    field(:organization_name, :string)
+    field(:application_id, :string)
+  end
+
+  @fields ~w(
+    access_token
+    hub_name
+    hub_color
+    organization_id
+    organization_name
+    organization_type
+    application_id
+  )a
+
+  def changeset(fly, attrs \\ %{}) do
+    fly
+    |> cast(attrs, @fields)
+    |> validate_required(@fields)
+    |> validate_color()
+    |> add_id()
+  end
+
+  defp validate_color(changeset) do
+    case get_field(changeset, :hub_color) do
+      nil ->
+        changeset
+
+      hex_color ->
+        if Utils.valid_hex_color?(hex_color) do
+          changeset
+        else
+          add_error(changeset, :hub_color, "not a valid color")
+        end
+    end
+  end
+
+  defp add_id(changeset) do
+    if application_id = get_field(changeset, :application_id) do
+      change(changeset, %{id: "fly-#{application_id}"})
+    else
+      validate_required(changeset, [:id])
```

Comment:
I don't think we need this because we already validate the `application_id` as required?